### PR TITLE
net: mDNS-SD

### PR DESCRIPTION
**Pull request by kim**
Adds mDNS discovery behind a feature flag.

Collaterals:

* When configured to listen on `0.0.0.0`, resolve the machine's network
  interfaces
* Instead of a single `SocketAddr`, return all addresses the peer is
  bound to (`Peer::listen_addrs`)
* Instead of specifiying the discovery in `PeerConfig`, supply it later
  to `Peer::bootstrap` (which is now public). This is necessary for
  mDNS, because it needs to know the bound addresses.
* Monomorphise `Static` disco, and define standard impls

### CODE REVIEW

File: ci/build-test [hunk 1]
```diff
@@ -2,7 +2,7 @@
 set -eou pipefail
 
 echo '--- Build'
-cargo build --verbose --workspace
+cargo build --verbose --workspace --all-features
```
Comment (xla) [inline ci/build-test:5]: Open the flood gates!

File: deny.toml [hunk 2]
```diff
@@ -204,6 +204,6 @@ allow-registry = ["https://github.com/rust-lang/crates.io-index"]
 # List of URLs for allowed Git repositories
 allow-git = [
     "https://github.com/djc/quinn",
-    "https://github.com/radicle-dev/radicle-keystore",
     "https://github.com/kim/ed25519-zebra",
+    "https://github.com/meilisearch/madness",
```
Comment (xla) [inline deny.toml:208]: I salute the boldness of this naming choice.

File: librad/Cargo.toml [hunk 3]
```diff
@@ -5,6 +5,10 @@ authors = ["The Radicle Team <dev@radicle.xyz>"]
 edition = "2018"
 license = "GPL-3.0-or-later"
 
+[features]
+default = []
+disco-mdns = ["mdns", "madness"]
```
Comment (xla) [inline librad/Cargo.toml:10]: Seems like the right way to introducing capabilities like this. :+1:

Comment (xla) [reply librad/Cargo.toml:10]: I feel like we missed an opportunity to call it: `disco-madness` :man_dancing:

File: librad/src/net/discovery.rs [hunk 4]
```diff
@@ -6,70 +6,113 @@
...
-
-    fn next(&mut self) -> Option<Self::Item> {
-        self.iter.find_map(|(peer_id, addrs)| {
-            // TODO: we might want to log resolver errors somewhere
-            addrs
-                .to_socket_addrs()
-                .ok()
-                .map(|resolved| (peer_id, resolved.collect()))
-        })
+// Better inference than `FromIterator`
```
Comment (xla) [inline librad/src/net/discovery.rs:74]: :facepalm:

File: librad/src/net/discovery/mdns.rs [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+                match packet {
+                    Packet::Query(queries) => {
+                        for query in queries {
+                            respond(&ptr, &cname, &listen_addrs, &mut srv, query)
+                        }
+                    },
+                    Packet::Response(resp) => {
+                        for info in discover(&cname, resp) {
+                            if tx.send(info).await.is_err() {
+                                tracing::info!("stopping the madness");
```
Comment (xla) [inline librad/src/net/discovery/mdns.rs:59]: :clap:

File: librad/src/net/discovery/mdns.rs [hunk 6]
```diff
@@ -0,0 +1,230 @@
...
+    use std::net::Ipv4Addr;
+
+    use crate::{keys::SecretKey, peer::PeerId};
+
+    // `madness` uses tokio, so no choice here -- MADNESS!
+    #[tokio::test]
+    // CI doesn't support multicast UDP
+    #[ignore]
+    async fn ohai() {
+        let lolek_id = PeerId::from(SecretKey::new());
```
Comment (xla) [inline librad/src/net/discovery/mdns.rs:210]: image

File: librad/src/net/peer.rs [hunk 7]
```diff
@@ -346,14 +332,16 @@ impl Peer {
         Ok((api, self.run_loop))
     }
 
-    async fn bootstrap<S, D>(config: PeerConfig<D, S>) -> Result<Self, BootstrapError>
+    pub async fn bootstrap<S, D>(config: PeerConfig<S>, disco: D) -> Result<Self, BootstrapError>
```
Comment (xla) [inline librad/src/net/peer.rs:335]: It makes a lot of sense now that it's in front of us, right there!

File: librad/src/net/peer.rs [hunk 8]
```diff
@@ -117,28 +117,14 @@ pub struct FetchInfo {
...
+pub struct PeerConfig<Signer> {
     pub signer: Signer,
     pub paths: Paths,
     pub listen_addr: SocketAddr,
     pub gossip_params: gossip::MembershipParams,
-    pub disco: Disco,
     pub storage_config: StorageConfig,
 }
 
-impl<D, S> PeerConfig<D, S>
```
Comment (xla) [inline librad/src/net/peer.rs:129]: Baiiii

**Review by xla — APPROVED** [review]
Looks very good! Doesn't seem to be as maddening as expected.

The streaming discos approach seems to hold so far.

🌮 🛄 📣 🍩

File: librad/Cargo.toml (hunk 3 above)
Comment (kim) [reply librad/Cargo.toml:10]: How would you feel about renaming `master` to `madness`?

Comment (xla) [reply librad/Cargo.toml:10]: I'd feel good about it.

**Comment by kim** [comment]
Not sure what the deal is with docker, it's working fine with `podman`.

File: librad/src/net/discovery/mdns.rs (hunk 6 above)
Comment (kim) [reply librad/src/net/discovery/mdns.rs:210]: I **need** to know the name of the pirate

**Comment by kim** [comment]
Closed via 4091e08c8cfc3ce291140badab28acbee1d6922a

File: librad/Cargo.toml (hunk 3 above)
Comment (FintanH) [reply librad/Cargo.toml:10]: > How would you feel about renaming master to madness?

➕

**Pull request closed without merging**